agent/local/state.go

```diff
@@ -0,0 +1,1031 @@
+	tokens *token.Store
+}
+
+// NewLocalState creates a new local state for the agent.
+func NewState(c Config, lg *log.Logger, tokens *token.Store) *State {
+	l := &State{
+		config:   c,
+		logger:   lg,
+		services: make(map[string]*ServiceState),
+		checks:   make(map[types.CheckID]*CheckState),
+		metadata: make(map[string]string),
+		tokens:   tokens,
+	}
+	l.SetDiscardCheckOutput(c.DiscardCheckOutput)
+	return l
+}
+
+// SetDiscardCheckOutput configures whether the check output
+// is discarded. This can be changed at runtime.
+func (l *State) SetDiscardCheckOutput(b bool) {
+	l.discardCheckOutput.Store(b)
+}
+
+// ServiceToken returns the configured ACL token for the given
+// service ID. If none is present, the agent's token is returned.
+func (l *State) ServiceToken(id string) string {
+	l.RLock()
+	defer l.RUnlock()
+	return l.serviceToken(id)
+}
+
+// serviceToken returns an ACL token associated with a service.
+// This method is not synchronized and the lock must already be held.
+func (l *State) serviceToken(id string) string {
+	var token string
+	if s := l.services[id]; s != nil {
+		token = s.Token
+	}
+	if token == "" {
+		token = l.tokens.UserToken()
+	}
+	return token
+}
+
+// AddService is used to add a service entry to the local state.
+// This entry is persistent and the agent will make a best effort to
+// ensure it is registered
+func (l *State) AddService(service *structs.NodeService, token string) error {
+	if service == nil {
+		return fmt.Errorf("no service")
+	}
+
+	// use the service name as id if the id was omitted
+	if service.ID == "" {
+		service.ID = service.Service
+	}
+
+	l.SetServiceState(&ServiceState{
+		Service: service,
+		Token:   token,
+	})
+	return nil
+}
+
+// RemoveService is used to remove a service entry from the local state.
+// The agent will make a best effort to ensure it is deregistered.
+func (l *State) RemoveService(id string) error {
+	l.Lock()
+	defer l.Unlock()
+
+	s := l.services[id]
+	if s == nil || s.Deleted {
+		return fmt.Errorf("Service %q does not exist", id)
+	}
+
+	// To remove the service on the server we need the token.
+	// Therefore, we mark the service as deleted and keep the
+	// entry around until it is actually removed.
+	s.InSync = false
+	s.Deleted = true
+	l.TriggerSyncChanges()
+
+	return nil
+}
+
+// Service returns the locally registered service that the
+// agent is aware of and are being kept in sync with the server
+func (l *State) Service(id string) *structs.NodeService {
+	l.RLock()
+	defer l.RUnlock()
+
+	s := l.services[id]
+	if s == nil || s.Deleted {
+		return nil
+	}
+	return s.Service
+}
+
+// Services returns the locally registered services that the
+// agent is aware of and are being kept in sync with the server
+func (l *State) Services() map[string]*structs.NodeService {
+	l.RLock()
+	defer l.RUnlock()
+
+	m := make(map[string]*structs.NodeService)
+	for id, s := range l.services {
+		if s.Deleted {
+			continue
+		}
+		m[id] = s.Service
```

Comment:
Should this be a `Clone()`?

Reply:
I suppose not since this is the immutable part.

Reply:
We only do `Clone` for `ServiceState` and `CheckState`.

Reply:
Also, `Clone` is shallow in that it will still return the pointer to the `NodeService`. See comment in line 50.